tests/test_downloadermiddleware_retry.py

```diff
@@ -104,5 +104,86 @@ def _test_retry_exception(self, req, exception):
         self.assertEqual(req, None)
 
 
+class MaxRetryTimesTest(unittest.TestCase):
+    def setUp(self):
+        self.crawler = get_crawler(Spider)
+        self.spider = self.crawler._create_spider('foo')
+        self.mw = RetryMiddleware.from_crawler(self.crawler)
+        self.mw.max_retry_times = 2
+
+    def test_without_metakey(self):
+        
+        req = Request('http://www.scrapytest.org/invalid_url')
+
+        # SETTINGS: RETRY_TIMES is NON-ZERO
```

Comment:
I believe you can split this into 3 different tests.
You can set `self.mw.max_retry_times` to a new value at the start of each test method. (or use a local `mw` each time).

Also, changing the middleware `max_retry_times` attribute while working on the same `req` object makes it hard (for me at least) to follow the expected behavior.
I think it's easier to test: 1) I have this middleware with these settings, + 1 request object with some meta (or not), 2) it goes through the middleware handling an exception for that request 3) check how many times it is being retried 4) end test